python/ray/util/spark/cluster_init.py

```diff
@@ -1054,15 +986,188 @@ def setup_ray_cluster(
             worker_node_options=worker_node_options,
             ray_temp_root_dir=ray_temp_root_dir,
             collect_log_to_path=collect_log_to_path,
+            autoscale=autoscale,
         )
-        cluster.wait_until_ready()  # NB: this line might raise error.
+
+        if not autoscale:
+            cluster.wait_until_ready()  # NB: this line might raise error.
 
         # If connect cluster successfully, set global _active_ray_cluster to be the
         # started cluster.
         _active_ray_cluster = cluster
     return cluster.address
 
 
+def _start_ray_worker_nodes(
+    *,
+    spark,
+    spark_job_group_id,
+    spark_job_group_desc,
+    num_worker_nodes,
+    using_stage_scheduling,
+    ray_head_ip,
+    ray_head_port,
+    ray_temp_dir,
+    num_cpus_per_node,
+    num_gpus_per_node,
+    heap_memory_per_node,
+    object_store_memory_per_node,
+    worker_node_options,
+    collect_log_to_path,
+    resources,
+):
+    # NB:
+    # In order to start ray worker nodes on spark cluster worker machines,
+    # We launch a background spark job:
+    #  1. Each spark task launches one ray worker node. This design ensures all ray
+    #     worker nodes have the same shape (same cpus / gpus / memory configuration).
+    #     If ray worker nodes have a non-uniform shape, the Ray cluster setup will
+    #     be non-deterministic and could create issues with node sizing.
+    #  2. A ray worker node is started via the `ray start` CLI. In each spark task,
+    #     a child process is started and will execute a `ray start ...` command in
+    #     blocking mode.
+    #  3. Each task will acquire a file lock for 10s to ensure that the ray worker
+    #     init will acquire a port connection to the ray head node that does not
+    #     contend with other worker processes on the same Spark worker node.
+    #  4. When the ray cluster is shutdown, killing ray worker nodes is implemented by:
+    #     Installing a PR_SET_PDEATHSIG signal for the `ray start ...` child processes
+    #     so that when parent process (pyspark task) is killed, the child processes
+    #     (`ray start ...` processes) will receive a SIGTERM signal, killing it.
+    #     Shutting down the ray cluster is performed by calling
+    #     `sparkContext.cancelJobGroup` to cancel the background spark job, sending a
+    #     SIGKILL signal to all spark tasks. Once the spark tasks are killed, this
+    #     triggers the sending of a SIGTERM to the child processes spawned by the
+    #     `ray_start ...` process.
+
+    # TODO: for autoscaler mode, wrap this mapper function to swallow exception,
+    #  to prevent spark task retries on failure.
```

Comment:
@jjyao 

To confirm behavior, assuming in autoscaling mode, a ray worker node crashes unexpectedly, will autoscaling framework trigger `node_provider.create_node` again to launch a new worker of the same shape ?

Reply:
It depends. If there is demand for that worker node or we are below min_workers, then a new worker will be launched by autoscaler.